File: mobile/lib/call.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:permission_handler/permission_handler.dart';
import 'call_service.dart';
import 'models.dart';

/// One voice/video call. Mirrors the desktop's design: 3 transceivers
/// (audio, camera-video, screen-video) negotiated up front + a 'ctrl' data
/// channel for live mute/camera status. Non-trickle ICE (gather then send SDP).
class CallSession extends ChangeNotifier {
  final String callId;
  final Peer peer;
  final bool isCaller;
  final String? offerSdp;
  final Future<bool> Function(String kind, String? sdp) sendSignal;
  final VoidCallback onClosed;

  RTCPeerConnection? _pc;
  RTCRtpTransceiver? _audioTx;
  RTCRtpTransceiver? _cameraTx;
  RTCRtpTransceiver? _screenTx;
  RTCDataChannel? _ctrl;
  MediaStream? _micStream;
  MediaStream? _camStream;
  MediaStream? _screenStream;
  MediaStreamTrack? _remoteAudioTrack;
  // Placeholder local streams associated with the camera/screen senders so the
  // outgoing SDP carries a real msid (stream id). Without it the peer receives
  // the track with an empty streams list and can't render it (renders black).
  MediaStream? _camLocalStream;
  MediaStream? _screenLocalStream;

  // remote media (bound to renderers in the UI)
  MediaStream? remoteCamera;
  MediaStream? remoteScreen;
  MediaStream? localCamera;
  MediaStream? localScreen;
  int _remoteVideoCount = 0;

  String status = 'connecting…';
  bool muted = false;
  bool deafened = false;
  bool camOn = false;
  bool screenOn = false;
  bool ended = false;
  bool connected = false;
  String camFacing = 'user'; // 'user' (front) or 'environment' (back)

  // remote live state (via ctrl channel)
  bool remoteMic = true;
  bool remoteCamOn = false;
  bool remoteScreenOn = false;
  bool remoteCamMirror = false; // peer's camera is mirrored (front) -> flip to correct it

  CallSession({
    required this.callId,
    required this.peer,
    required this.isCaller,
    required this.offerSdp,
    required this.sendSignal,
    required this.onClosed,
  });

  Future<void> start() async {
    try {
      await _createPc();
      if (isCaller) {
        await _makeOffer();
      } else {
        await _makeAnswer(offerSdp!);
      }
    } catch (e) {
      _setStatus('Call failed: $e');
      end(notifyPeer: false);
    }
  }

  Future<void> _createPc() async {
    _pc = await createPeerConnection({
      'iceServers': [
        {'urls': 'stun:stun.l.google.com:19302'},
      ],
    });

    _pc!.onTrack = (RTCTrackEvent e) {
      final track = e.track;
      if (track.kind == 'audio') {
        // remote audio plays automatically; keep a ref so Deafen can mute it
        _remoteAudioTrack = track;
        track.enabled = !deafened;
        return;
      }
      final stream = e.streams.isNotEmpty ? e.streams.first : null;
      if (stream == null) return;
      // Distinguish the two remote video m-lines by their transceiver mid
      // (audio=0, camera=1, screen=2 — the order we add them).
      final mid = e.transceiver?.mid;
      final isScreen = mid == '2' || (mid == null && _remoteVideoCount > 0);
      _remoteVideoCount++;
      if (isScreen) {
        remoteScreen = stream;
      } else {
        remoteCamera = stream;
      }
      notifyListeners();
    };

    _pc!.onConnectionState = (state) {
      if (state == RTCPeerConnectionState.RTCPeerConnectionStateConnected) {
        _onConnected();
      } else if (state == RTCPeerConnectionState.RTCPeerConnectionStateFailed) {
        _setStatus('Connection failed');
      } else if (state == RTCPeerConnectionState.RTCPeerConnectionStateDisconnected) {
        if (connected) _setStatus('Reconnecting…');
      }
    };

    // ICE state is more reliable than peer-connection state on some devices, so
    // settle the status from whichever fires first.
    _pc!.onIceConnectionState = (state) {
      if (state == RTCIceConnectionState.RTCIceConnectionStateConnected ||
          state == RTCIceConnectionState.RTCIceConnectionStateCompleted) {
        _onConnected();
      }
    };

    if (isCaller) {
      _ctrl = await _pc!.createDataChannel('ctrl', RTCDataChannelInit());
      _bindCtrl(_ctrl!);
    } else {
      _pc!.onDataChannel = (channel) {
        if (channel.label == 'ctrl') {
          _ctrl = channel;
          _bindCtrl(channel);
        }
      };
    }
  }

  Future<void> _makeOffer() async {
    final mic = await _getMic();
    _audioTx = await _pc!.addTransceiver(
      kind: RTCRtpMediaType.RTCRtpMediaTypeAudio,
      init: RTCRtpTransceiverInit(direction: TransceiverDirection.SendRecv),
    );
    if (mic != null) await _audioTx!.sender.replaceTrack(mic);
    _camLocalStream = await createLocalMediaStream('filedrop-cam');
    _screenLocalStream = await createLocalMediaStream('filedrop-screen');
    _cameraTx = await _pc!.addTransceiver(
      kind: RTCRtpMediaType.RTCRtpMediaTypeVideo,
      init: RTCRtpTransceiverInit(direction: TransceiverDirection.SendRecv, streams: [_camLocalStream!]),
    );
    _screenTx = await _pc!.addTransceiver(
      kind: RTCRtpMediaType.RTCRtpMediaTypeVideo,
      init: RTCRtpTransceiverInit(direction: TransceiverDirection.SendRecv, streams: [_screenLocalStream!]),
    );

    final offer = await _pc!.createOffer();
    await _pc!.setLocalDescription(offer);
    _setStatus('Calling…');
    await _gather();
    final desc = await _pc!.getLocalDescription();
    final ok = await sendSignal('offer', desc?.sdp);
    if (!connected) _setStatus(ok ? 'Ringing…' : "Couldn't reach them");
  }

  Future<void> _makeAnswer(String sdp) async {
    _setStatus('Connecting…');
    await _pc!.setRemoteDescription(RTCSessionDescription(sdp, 'offer'));
    await _assignTransceivers();
    // Associate a stream with our outgoing camera/screen senders so the caller
    // receives them with a real msid and can render them (else: black/no video).
    _camLocalStream = await createLocalMediaStream('filedrop-cam');
    _screenLocalStream = await createLocalMediaStream('filedrop-screen');
    await _cameraTx?.sender.setStreams([_camLocalStream!]);
    await _screenTx?.sender.setStreams([_screenLocalStream!]);
    final mic = await _getMic();
    if (mic != null && _audioTx != null) await _audioTx!.sender.replaceTrack(mic);
    await _audioTx?.setDirection(TransceiverDirection.SendRecv);
    await _cameraTx?.setDirection(TransceiverDirection.SendRecv);
    await _screenTx?.setDirection(TransceiverDirection.SendRecv);

    final answer = await _pc!.createAnswer();
    await _pc!.setLocalDescription(answer);
    await _gather();
    final desc = await _pc!.getLocalDescription();
    final ok = await sendSignal('answer', desc?.sdp);
    if (!connected && !ok) _setStatus("Couldn't reach them");
  }

  Future<void> _assignTransceivers() async {
    final txs = await _pc!.getTransceivers();
    final audios = txs.where((t) => t.receiver.track?.kind == 'audio').toList();
    final videos = txs.where((t) => t.receiver.track?.kind == 'video').toList();
    _audioTx = audios.isNotEmpty ? audios.first : null;
    _cameraTx = videos.isNotEmpty ? videos.first : null;
    _screenTx = videos.length > 1 ? videos[1] : null;
  }

  /// Incoming answer/decline/busy/hangup routed from the service.
  Future<void> handleSignal(String kind, String? sdp) async {
    if (ended) return;
    if (kind == 'answer' && sdp != null) {
      await _pc!.setRemoteDescription(RTCSessionDescription(sdp, 'answer'));
    } else if (kind == 'decline') {
      end(message: 'Call declined', notifyPeer: false);
    } else if (kind == 'busy') {
      end(message: 'Busy — try later', notifyPeer: false);
    } else if (kind == 'hangup') {
      end(message: 'Call ended', notifyPeer: false);
    }
  }

  Future<MediaStreamTrack?> _getMic() async {
    if (_micStream != null) return _micStream!.getAudioTracks().first;
    try {
      await Permission.microphone.request();
      _micStream = await navigator.mediaDevices.getUserMedia({
        'audio': {
          'echoCancellation': true,
          'noiseSuppression': true, // voice suppression
          'autoGainControl': true,
        },
        'video': false,
      });
      return _micStream!.getAudioTracks().first;
    } catch (_) {
      return null;
    }
  }

  Future<void> toggleMute() async {
    muted = !muted;
    _micStream?.getAudioTracks().forEach((t) => t.enabled = !muted);
    _sendCtrl();
    notifyListeners();
  }

  Future<void> toggleCamera() async {
    if (camOn) {
      _camStream?.getTracks().forEach((t) => t.stop());
      _camStream = null;
      localCamera = null;
      camOn = false;
      if (_cameraTx != null) await _cameraTx!.sender.replaceTrack(null);
    } else {
      try {
        await Permission.camera.request();
        _camStream = await navigator.mediaDevices.getUserMedia({
          'audio': false,
          'video': {
            'facingMode': 'user',
            'width': {'ideal': 1920},
            'height': {'ideal': 1080},
            'frameRate': {'ideal': 30},
          },
        });
        final track = _camStream!.getVideoTracks().first;
        if (_cameraTx != null) {
          await _cameraTx!.sender.replaceTrack(track);
          await _setMaxBitrate(_cameraTx!.sender, 3000000); // ~3 Mbps for crisp video
        }
        localCamera = _camStream;
        camOn = true;
        camFacing = 'user';
      } catch (_) {
        return;
      }
    }
    _sendCtrl();
    notifyListeners();
  }

  /// Flip between the front and back camera mid-call.
  Future<void> switchCamera() async {
    if (!camOn || _camStream == null) return;
    final tracks = _camStream!.getVideoTracks();
    if (tracks.isEmpty) return;
    try {
      await Helper.switchCamera(tracks.first);
      camFacing = camFacing == 'user' ? 'environment' : 'user';
      _sendCtrl(); // tell the peer whether to flip our video (front is mirrored)
      notifyListeners();
    } catch (_) {}
  }

  Future<void> toggleDeafen() async {
    deafened = !deafened;
    _remoteAudioTrack?.enabled = !deafened;
    if (deafened && !muted) {
      await toggleMute(); // deafening also mutes your mic, like the desktop
    }
    notifyListeners();
  }

  Future<void> toggleScreen() async {
    if (screenOn) {
      _screenStream?.getTracks().forEach((t) => t.stop());
      _screenStream = null;
      localScreen = null;
      screenOn = false;
      if (_screenTx != null) await _screenTx!.sender.replaceTrack(null);
    } else {
      try {
        _screenStream = await navigator.mediaDevices.getDisplayMedia({
          'video': {'frameRate': {'ideal': 30}},
          'audio': false,
        });
        final track = _screenStream!.getVideoTracks().first;
        if (_screenTx != null) {
          await _screenTx!.sender.replaceTrack(track);
          await _setMaxBitrate(_screenTx!.sender, 4000000); // ~4 Mbps so screen text stays sharp
        }
        localScreen = _screenStream;
        screenOn = true;
        // stopped from the system "casting" notification
        track.onEnded = () {
          if (screenOn) toggleScreen();
        };
      } catch (_) {
        return;
      }
    }
    _sendCtrl();
    notifyListeners();
  }

  void _bindCtrl(RTCDataChannel ch) {
    ch.onDataChannelState = (state) {
      if (state == RTCDataChannelState.RTCDataChannelOpen) _sendCtrl();
    };
    ch.onMessage = (msg) {
      try {
        final s = msg.text;
        // simple csv-free JSON
        final m = _decode(s);
        remoteMic = m['mic'] ?? remoteMic;
        remoteCamOn = m['cam'] ?? remoteCamOn;
        remoteScreenOn = m['screen'] ?? remoteScreenOn;
        remoteCamMirror = m['mir'] ?? remoteCamMirror;
        notifyListeners();
      } catch (_) {}
    };
  }

  void _sendCtrl() {
    if (_ctrl == null) return;
    try {
      _ctrl!.send(RTCDataChannelMessage(_encode({
        'mic': !muted && _micStream != null,
        'cam': camOn,
        'screen': screenOn,
        'mir': camOn && camFacing == 'user', // front camera is mirrored
      })));
    } catch (_) {}
  }

  Future<void> _gather() async {
    if (_pc!.iceGatheringState == RTCIceGatheringState.RTCIceGatheringStateComplete) return;
    final c = Completer<void>();
    Timer? timer;
    _pc!.onIceGatheringState = (s) {
      if (s == RTCIceGatheringState.RTCIceGatheringStateComplete && !c.isCompleted) c.complete();
    };
    timer = Timer(const Duration(milliseconds: 2500), () {
      if (!c.isCompleted) c.complete();
    });
    await c.future;
    timer.cancel();
  }

  void _onConnected() {
    if (connected) {
      _sendCtrl();
      return;
    }
    connected = true;
    _setStatus('In call');
    _sendCtrl(); // resync mute/camera/screen state now that we're up
    _startCallService();
  }

  /// Start the ongoing-call foreground service so the call survives the app
  /// being backgrounded/closed (HyperOS/MIUI kill background apps otherwise).
  Future<void> _startCallService() async {
    try {
      await Permission.notification.request();
    } catch (_) {}
    // One-tap "let it run in the background" — the same exemption WhatsApp/Discord
    // ask for, so calls survive backgrounding without locking the app in recents.
    try {
      if (await Permission.ignoreBatteryOptimizations.isDenied) {
        await Permission.ignoreBatteryOptimizations.request();
      }
    } catch (_) {}
    await startCallService(text: 'In call with ${peer.name}');
  }

  void _setStatus(String s) {
    status = s;
    notifyListeners();
  }

  /// Raise the per-sender encoding bitrate so video looks crisp on good links.
  Future<void> _setMaxBitrate(RTCRtpSender sender, int bps) async {
    try {
      final params = sender.parameters;
      if (params.encodings == null || params.encodings!.isEmpty) {
        params.encodings = [RTCRtpEncoding(maxBitrate: bps)];
      } else {
        for (final enc in params.encodings!) {
          enc.maxBitrate = bps;
        }
      }
      await sender.setParameters(params);
    } catch (_) {}
  }

  /// Hang up. notifyPeer=true sends a hangup to the other side.
  Future<void> end({String? message, bool notifyPeer = true}) async {
    if (ended) return;
    ended = true;
    stopCallService(); // tear down the ongoing-call notification + keep-alive
    if (message != null) status = message;
    if (notifyPeer) sendSignal('hangup', null);
    for (final s in [_micStream, _camStream, _screenStream, _camLocalStream, _screenLocalStream]) {
      s?.getTracks().forEach((t) => t.stop());
      await s?.dispose();
    }
    _micStream = null;
    _camStream = null;
    _screenStream = null;
    _camLocalStream = null;
    _screenLocalStream = null;
    try {
      await _ctrl?.close();
    } catch (_) {}
    try {
      await _pc?.close();
    } catch (_) {}
    _pc = null;
    notifyListeners();
    onClosed();
  }

  // tiny JSON helpers for the ctrl channel
  String _encode(Map<String, bool> m) =>
      '{"mic":${m['mic']},"cam":${m['cam']},"screen":${m['screen']},"mir":${m['mir']}}';
  Map<String, bool> _decode(String s) {
    bool g(String k) => RegExp('"$k"\\s*:\\s*true').hasMatch(s);
    return {'mic': g('mic'), 'cam': g('cam'), 'screen': g('screen'), 'mir': g('mir')};
  }
}

File: src/renderer/call.js
'use strict';

const api = window.filedrop;

// quality presets for screen share
const QUALITY = { '720': { w: 1280, h: 720 }, '1080': { w: 1920, h: 1080 }, source: null };
const BITRATE = { '720': 2500000, '1080': 5000000, source: 8000000 };

let pc = null;
let init = null; // { role, callId, peer, offerSdp }
const tx = { audio: null, camera: null, screen: null };
let ctrl = null;

let micStream = null;
let camStream = null;
let screenStream = null;
let currentSourceId = null;

let muted = false;
let deafened = false;
let camOn = false;
let screenOn = false;
let ended = false;

let remoteVideoCount = 0;
const remote = { mic: true, cam: false, screen: false, deaf: false };

// elements
const el = (id) => document.getElementById(id);
const remoteAudio = el('remote-audio');
const remoteCamera = el('remote-camera');
const remoteScreen = el('remote-screen');
const localCamera = el('local-camera');
const localScreen = el('local-screen');

// ---------------------------------------------------------------------------
boot();

async function boot() {
  wireControls();
  api.onCallSignal(handleSignal);
  api.onCallInit((i) => start(i)); // window reuse
  const pulled = await api.callReady();
  if (pulled) start(pulled);
}

async function start(i) {
  if (pc) cleanupPc();
  ended = false;
  init = i;
  el('peer-name').textContent = (i.peer && i.peer.name) || 'Call';
  setStatus(i.role === 'caller' ? 'calling…' : 'connecting…');
  createPc();
  try {
    if (i.role === 'caller') await makeOffer();
    else await makeAnswer(i.offerSdp);
  } catch (err) {
    toast('Could not start call: ' + err.message);
    end('failed', 'failed');
  }
}

// ---------------------------------------------------------------------------
// peer connection
// ---------------------------------------------------------------------------
function createPc() {
  pc = new RTCPeerConnection({ iceServers: [] }); // LAN only — host candidates

  pc.ontrack = (e) => {
    const track = e.track;
    if (track.kind === 'audio') {
      remoteAudio.srcObject = new MediaStream([track]);
      remoteAudio.muted = deafened;
    } else {
      const idx = remoteVideoCount++;
      const target = idx === 0 ? remoteCamera : remoteScreen;
      target.srcObject = new MediaStream([track]);
      // reveal as a fallback if ctrl state hasn't arrived
      track.onunmute = () => {
        if (idx === 0) remote.cam = true;
        else remote.screen = true;
        layoutRemote();
      };
      track.onmute = () => {
        if (idx === 0) remote.cam = false;
        else remote.screen = false;
        layoutRemote();
      };
    }
  };

  pc.onconnectionstatechange = () => {
    const s = pc.connectionState;
    if (s === 'connected') setStatus('connected', 'connected');
    else if (s === 'failed') setStatus('connection failed', 'failed');
    else if (s === 'disconnected') setStatus('reconnecting…');
    else if (s === 'closed') {/* handled by end() */}
  };

  if (init.role === 'caller') {
    ctrl = pc.createDataChannel('ctrl');
    bindCtrl(ctrl);
  } else {
    pc.ondatachannel = (e) => {
      if (e.channel.label === 'ctrl') {
        ctrl = e.channel;
        bindCtrl(ctrl);
      }
    };
  }
}

async function makeOffer() {
  const mic = await getMic();
  tx.audio = pc.addTransceiver(mic || 'audio', { direction: 'sendrecv' });
  tx.camera = pc.addTransceiver('video', { direction: 'sendrecv' });
  tx.screen = pc.addTransceiver('video', { direction: 'sendrecv' });

  const offer = await pc.createOffer();
  await pc.setLocalDescription(offer);
  await gather(pc);
  api.rtcOut({ kind: 'offer', sdp: pc.localDescription.sdp });
  setStatus('ringing…');
}

async function makeAnswer(offerSdp) {
  await pc.setRemoteDescription({ type: 'offer', sdp: offerSdp });
  assignTransceivers();
  const mic = await getMic();
  if (mic && tx.audio) await tx.audio.sender.replaceTrack(mic);
  // make every line bidirectional so we can enable cam/screen later w/o reneg
  for (const t of [tx.audio, tx.camera, tx.screen]) if (t) t.direction = 'sendrecv';

  const answer = await pc.createAnswer();
  await pc.setLocalDescription(answer);
  await gather(pc);
  api.rtcOut({ kind: 'answer', sdp: pc.localDescription.sdp });
}

function assignTransceivers() {
  const txs = pc.getTransceivers();
  tx.audio = txs.find((t) => t.receiver.track && t.receiver.track.kind === 'audio') || null;
  const vids = txs.filter((t) => t.receiver.track && t.receiver.track.kind === 'video');
  tx.camera = vids[0] || null;
  tx.screen = vids[1] || null;
}

function handleSignal(sig) {
  if (!sig || ended) return;
  if (sig.kind === 'answer') {
    pc.setRemoteDescription({ type: 'answer', sdp: sig.sdp }).catch((e) => toast('answer error: ' + e.message));
  } else if (sig.kind === 'decline') {
    end('Call declined', 'ended');
  } else if (sig.kind === 'busy') {
    end('Busy — try again later', 'ended');
  } else if (sig.kind === 'hangup') {
    end('Call ended', 'ended');
  } else if (sig.kind === 'timeout') {
    end('No answer', 'ended');
  } else if (sig.kind === 'unreachable') {
    end('Device unreachable', 'ended');
  }
}

/** Wait until ICE gathering is complete (or a short timeout). */
function gather(peer) {
  if (peer.iceGatheringState === 'complete') return Promise.resolve();
  return new Promise((resolve) => {
    const done = () => {
      peer.removeEventListener('icegatheringstatechange', check);
      clearTimeout(t);
      resolve();
    };
    const check = () => {
      if (peer.iceGatheringState === 'complete') done();
    };
    peer.addEventListener('icegatheringstatechange', check);
    const t = setTimeout(done, 2500); // LAN gathers fast; cap the wait
  });
}

// ---------------------------------------------------------------------------
// control channel (live mute/cam/screen status)
// ---------------------------------------------------------------------------
function bindCtrl(ch) {
  ch.onopen = () => sendCtrl();
  ch.onmessage = (e) => {
    try {
      const s = JSON.parse(e.data);
      Object.assign(remote, s);
      layoutRemote();
      renderBadges();
    } catch (_) {}
  };
}

function sendCtrl() {
  if (!ctrl || ctrl.readyState !== 'open') return;
  try {
    ctrl.send(JSON.stringify({ mic: !muted && !!micStream, cam: camOn, screen: screenOn, deaf: deafened }));
  } catch (_) {}
}

// ---------------------------------------------------------------------------
// media helpers
// ---------------------------------------------------------------------------
async function getMic() {
  if (micStream) return micStream.getAudioTracks()[0];
  try {
    micStream = await navigator.mediaDevices.getUserMedia({
      audio: { echoCancellation: true, noiseSuppression: true, autoGainControl: true },
    });
    return micStream.getAudioTracks()[0];
  } catch (err) {
    toast('Microphone unavailable — you can still hear the call');
    return null;
  }
}

async function toggleCamera() {
  if (camOn) {
    if (camStream) camStream.getTracks().forEach((t) => t.stop());
    camStream = null;
    camOn = false;
    if (tx.camera) await tx.camera.sender.replaceTrack(null);
    localCamera.classList.add('hidden');
    localCamera.srcObject = null;
  } else {
    try {
      camStream = await navigator.mediaDevices.getUserMedia({
        video: { width: { ideal: 1280 }, height: { ideal: 720 }, frameRate: { ideal: 30 } },
      });
      const track = camStream.getVideoTracks()[0];
      if (tx.camera) {
        await tx.camera.sender.replaceTrack(track);
        // cap the camera encoder so it can't hog the link (screen has its own cap)
        try {
          const p = tx.camera.sender.getParameters();
          if (!p.encodings || !p.encodings.length) p.encodings = [{}];
          p.encodings[0].maxBitrate = 1500000; // ~1.5 Mbps for 720p
          p.encodings[0].maxFramerate = 30;
          await tx.camera.sender.setParameters(p);
        } catch (_) {}
      }
      localCamera.srcObject = camStream;
      localCamera.classList.remove('hidden');
      camOn = true;
    } catch (err) {
      toast('Camera unavailable: ' + err.message);
      return;
    }
  }
  el('btn-camera').classList.toggle('active', camOn);
  sendCtrl();
}

async function startScreen(sourceId, fps, quality) {
  try {
    const q = QUALITY[quality];
    const mandatory = { chromeMediaSource: 'desktop', chromeMediaSourceId: sourceId, maxFrameRate: Number(fps) };
    if (q) {
      mandatory.maxWidth = q.w;
      mandatory.maxHeight = q.h;
    }
    const stream = await navigator.mediaDevices.getUserMedia({ audio: false, video: { mandatory } });
    if (screenStream) screenStream.getTracks().forEach((t) => t.stop());
    screenStream = stream;
    currentSourceId = sourceId;
    const track = stream.getVideoTracks()[0];
    track.onended = () => stopScreen(); // user stopped via OS overlay
    if (tx.screen) {
      await tx.screen.sender.replaceTrack(track);
      await applyScreenParams(fps, quality);
    }
    localScreen.srcObject = stream;
    localScreen.classList.remove('hidden');
    screenOn = true;
    el('btn-screen').classList.add('active');
    el('live-quality').classList.remove('hidden');
    sendCtrl();
  } catch (err) {
    toast('Screen share failed: ' + err.message);
  }
}

async function applyScreenParams(fps, quality) {
  if (!tx.screen || !tx.screen.sender) return;
  try {
    const p = tx.screen.sender.getParameters();
    if (!p.encodings || !p.encodings.length) p.encodings = [{}];
    p.encodings[0].maxBitrate = BITRATE[quality] || 5000000;
    p.encodings[0].maxFramerate = Number(fps);
    await tx.screen.sender.setParameters(p);
  } catch (_) {}
}

async function stopScreen() {
  if (!screenOn) return;
  if (screenStream) screenStream.getTracks().forEach((t) => t.stop());
  screenStream = null;
  currentSourceId = null;
  if (tx.screen) await tx.screen.sender.replaceTrack(null);
  localScreen.classList.add('hidden');
  localScreen.srcObject = null;
  screenOn = false;
  el('btn-screen').classList.remove('active');
  el('live-quality').classList.add('hidden');
  sendCtrl();
}

function toggleMute() {
  muted = !muted;
  if (micStream) micStream.getAudioTracks().forEach((t) => (t.enabled = !muted));
  el('btn-mic').classList.toggle('danger-on', muted);
  el('btn-mic').querySelector('.lbl').textContent = muted ? 'Unmute' : 'Mute';
  sendCtrl();
}

function toggleDeafen() {
  deafened = !deafened;
  remoteAudio.muted = deafened;
  // deafening also mutes your mic (like Discord)
  if (deafened && !muted) toggleMute();
  el('btn-deafen').classList.toggle('danger-on', deafened);
  el('btn-deafen').querySelector('.lbl').textContent = deafened ? 'Undeafen' : 'Deafen';
  sendCtrl();
}

// ---------------------------------------------------------------------------
// remote layout + badges
// ---------------------------------------------------------------------------
function layoutRemote() {
  const showScreen = remote.screen && remoteScreen.srcObject;
  const showCam = remote.cam && remoteCamera.srcObject;
  remoteScreen.classList.toggle('hidden', !showScreen);
  remoteCamera.classList.toggle('hidden', !showCam);
  // if both, camera goes to a corner; if only camera, it fills
  remoteCamera.classList.toggle('cornered', showScreen && showCam);
  el('placeholder').classList.toggle('hidden', showScreen || showCam);
  renderBadges();
}

function renderBadges() {
  const b = [];
  if (!remote.mic || remote.deaf) b.push('<span class="on">🔇</span>');
  if (remote.cam) b.push('<span class="on">📷</span>');
  if (remote.screen) b.push('<span class="on">🖥️</span>');
  el('remote-badges').innerHTML = b.join(' ');
}

// ---------------------------------------------------------------------------
// screen-share picker
// ---------------------------------------------------------------------------
async function openSharePicker() {
  el('share-modal').classList.remove('hidden');
  const box = el('share-sources');
  box.textContent = 'Loading sources…';
  const sources = await api.desktopSources();
  if (!sources.length) {
    box.textContent = 'No screens/windows available.';
    return;
  }
  box.innerHTML = '';
  for (const s of sources) {
    const node = document.createElement('div');
    node.className = 'source';
    node.innerHTML =
      `<img src="${s.thumbnail || ''}" alt="">` +
      `<div class="src-name">${escapeHtml(s.name)}</div>`;
    node.onclick = () => {
      el('share-modal').classList.add('hidden');
      startScreen(s.id, el('pick-fps').value, el('pick-q').value);
      // sync live selectors with the picked values
      el('live-fps').value = el('pick-fps').value;
      el('live-q').value = el('pick-q').value;
    };
    box.appendChild(node);
  }
}

// ---------------------------------------------------------------------------
// wiring
// ---------------------------------------------------------------------------
function wireControls() {
  el('btn-mic').onclick = toggleMute;
  el('btn-deafen').onclick = toggleDeafen;
  el('btn-camera').onclick = toggleCamera;
  el('btn-screen').onclick = () => (screenOn ? stopScreen() : openSharePicker());
  el('btn-hangup').onclick = () => end('Call ended', 'ended', true);
  el('share-cancel').onclick = () => el('share-modal').classList.add('hidden');

  const liveChange = () => {
    if (screenOn && currentSourceId) startScreen(currentSourceId, el('live-fps').value, el('live-q').value);
  };
  el('live-fps').onchange = liveChange;
  el('live-q').onchange = liveChange;

  window.addEventListener('beforeunload', () => {
    if (!ended) {
      try { api.callEnded(); } catch (_) {}
    }
  });
}

// ---------------------------------------------------------------------------
function end(message, statusClass, userHangup) {
  if (ended) {
    if (userHangup) window.close();
    return;
  }
  ended = true;
  setStatus(message, statusClass);
  cleanupPc();
  if (userHangup) {
    window.close(); // main posts hangup on window 'closed'
  } else {
    toast(message);
    setTimeout(() => window.close(), 1200);
  }
}

function cleanupPc() {
  for (const s of [micStream, camStream, screenStream]) {
    if (s) s.getTracks().forEach((t) => t.stop());
  }
  micStream = camStream = screenStream = null;
  try {
    if (ctrl) ctrl.close();
  } catch (_) {}
  try {
    if (pc) pc.close();
  } catch (_) {}
  pc = null;
}

function setStatus(text, cls) {
  const s = el('status');
  s.textContent = text;
  s.className = 'status' + (cls ? ' ' + cls : '');
}

let toastTimer = null;
function toast(msg) {
  const t = el('toast');
  t.textContent = msg;
  t.classList.remove('hidden');
  clearTimeout(toastTimer);
  toastTimer = setTimeout(() => t.classList.add('hidden'), 3500);
}

function escapeHtml(s) {
  return String(s).replace(/[&<>"']/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]));
}
